ui/src/app/utils/cluster-initializer.tsx

```diff
@@ -0,0 +1,301 @@
+    }
+    await clusterApi().deleteCluster(this.workspaceNamespace, {signal: this.abortSignal});
+    this.deleteCount++;
+  }
+
+  private isClusterRunning(): boolean {
+    return this.currentCluster && this.currentCluster.status === ClusterStatus.Running;
+  }
+
+  private isClusterStopped(): boolean {
+    return this.currentCluster && this.currentCluster.status === ClusterStatus.Stopped;
+  }
+
+  private isClusterErrored(): boolean {
+    return this.currentCluster && this.currentCluster.status === ClusterStatus.Error;
+  }
+
+  private isNotFoundError(e: any): boolean {
+    // Our Swagger-generated APIs throw an error of type Response on a non-success status code.
+    return e instanceof Response && e.status === 404;
+  }
+
+  private handleUnknownError(e: any) {
+    if (e instanceof Response && e.status >= 500 && e.status < 600) {
+      this.serverErrorCount++;
+    }
+    reportError(e);
+  }
+
+  private tooManyServerErrors(): boolean {
+    return this.serverErrorCount > this.maxServerErrorCount;
+  }
+
+  /**
+   * Runs the cluster intiailizer flow.
+   *
+   * The strategy here is to poll the getCluster endpoint for cluster status, waiting for the
+   * cluster to reach the ready state (ClusterStatus.Running) or an error state which can be
+   * recovered from. Action will be taken where possible: a stopped cluster will trigger a call to
+   * startCluster, a nonexistent cluster will trigger a call to createCluster, and an errored
+   * cluster will trigger a call to deleteCluster in an attempt to retry cluster creation.
+   *
+   * @return A Promise which resolves with a Cluster or rejects with a
+   * ClusterInitializationFailedError, which holds a message and the current Cluster object (if one
+   * existed at the time of failure).
+   */
+  public async run(): Promise<Cluster> {
+    if (this.isInitializing) {
+      throw new Error('Initialization is already in progress');
+    }
+    this.initializeStartTime = Date.now();
+
+    return new Promise((resolve, reject) => {
+      this.resolve = resolve;
+      this.reject = reject as (error: Error) => {};
+      this.poll();
+    }) as Promise<Cluster>;
+  }
+
+  private async poll() {
+    // Overall strategy: continue polling the get-cluster endpoint, with capped exponential backoff,
+    // until we either reach our goal state (a RUNNING cluster) or run up against the overall
+    // timeout threshold.
+    //
+    // Certain cluster states require active intervention, such as deleting or resuming the cluster;
+    // these are handled within the the polling loop.
+    if (this.abortSignal && this.abortSignal.aborted) {
+      // We'll bail out early if an abort signal was triggered while waiting for the poll cycle.
+      return this.reject(
+        new ClusterInitializationFailedError('Request was aborted.', this.currentCluster));
+    }
+    if (Date.now() - this.initializeStartTime > this.overallTimeout) {
```

Comment:
nit: breaking this chunk out into a `checkTimeout` method could make `poll()` shorter and raise its abstraction level a bit.